src/Text/Pandoc/Lua/PreorderTraversal.hs

```diff
@@ -0,0 +1,122 @@
+{-# LANGUAGE FlexibleContexts      #-}
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE LambdaCase           #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{- |
+Module      : Text.Pandoc.Lua.PreorderTraversal
+Copyright   : © 2012-2021 John MacFarlane,
+              © 2017-2021 Albert Krewinkel
+License     : GNU GPL, version 2 or above
+Maintainer  : Albert Krewinkel <tarleb+pandoc@moltkeplatz.de>
+
+Walk documents in a filter-suitable way, in regular order.
+-}
+module Text.Pandoc.Lua.PreorderTraversal
+  ( PreorderTraversalNode (..)
+  , Topdown (..)
+  , TraversalControl (..)
+  )
+where
+
+import Control.Monad ((>=>))
+import Text.Pandoc.Definition
+import Text.Pandoc.Lua.Walk
+import Text.Pandoc.Walk
+
+-- | Helper type to do a preorder traversal of a subtree.
+data PreorderTraversalNode
+  = PTBlock Block
+  | PTBlocks [Block]
+  | PTInline Inline
+  | PTInlines [Inline]
+
+data Topdown = Topdown
+  { topdownControl :: TraversalControl
+  , topdownNode :: PreorderTraversalNode
+  }
+
+tdInlines :: PreorderTraversalNode -> [Inline]
+tdInlines = \case
+  PTInlines xs -> xs
+  PTInline x   -> [x]
+  _            -> error $ "The 'impossible' has happened."
+                       ++ "Please report this as a bug"
+
+tdBlocks :: PreorderTraversalNode -> [Block]
+tdBlocks = \case
+  PTBlocks xs  -> xs
+  PTBlock x    -> [x]
+  PTInlines xs -> [Plain xs]
+  PTInline x   -> [Plain [x]]
+
+instance {-# OVERLAPPING #-} Walkable Topdown [Block] where
+  walkM f = f . Topdown Continue . PTBlocks >=> \case
+    Topdown Stop     blksNode -> return $ tdBlocks blksNode
+    Topdown Continue blksNode -> mconcat <$>
+      traverse (f . Topdown Continue . PTBlock >=> \case
+                   Topdown Stop     node -> return $ tdBlocks node
+                   Topdown Continue node -> traverse (walkBlockM f) $
+                                            tdBlocks node)
+               (tdBlocks blksNode)
+  query f blks = f (Topdown Continue $ PTBlocks blks)
+              <> mconcat (map (queryBlock f) blks)
+
+instance {-# OVERLAPPING #-} Walkable Topdown [Inline] where
+  walkM f = f . Topdown Continue . PTInlines >=> \case
+    Topdown Stop     node -> return $ tdInlines node
+    Topdown Continue node -> mconcat <$>
+      traverse (f . Topdown Continue . PTInline >=> \case
+                   Topdown Stop     node' -> return $ tdInlines node'
+                   Topdown Continue node' -> traverse (walkInlineM f) $
+                                             tdInlines node')
+               (tdInlines node)
+  query f inlns = f (Topdown Continue $ PTInlines inlns)
```

Comment:
Why are these implemented for [Block] and [Inline] but not Block and Inline?

Reply:
Primarily because we must assume `f` to return a (wrapped) list, not a single element. So we can't implement it for Block and Inline in a sensible way -- or wouldn't use it, anyway.

The code is probably a slightly hacky misuse of the `Walkable` class, but that's already true for the current Lua filter implementation.